primitives/src/runtime_api.rs

```diff
@@ -110,7 +110,7 @@
 //! All staging API functions should use primitives from `vstaging`. They should be clearly separated
 //! from the stable primitives.
 
-use crate::v2;
+use crate::{v2, vstaging};
```

Comment:
nit: I know this is still a draft, but I have some questions anyway. :) Why is the new module called `vstaging`?

Reply:
Well, I agree that's not too clear, but that's how it is. "v2" for "version two", "v3" for "version three", "vstaging" for "version staging". Not my invention, I don't quite understand it either :)

Still need some advice from @tdimitrov if it's placed correctly.

Reply:
Yes, this is good.

@m-cat the node (aka the client, the native  code) communicates with the runtime (which is wasm) via so called [runtime api](https://paritytech.github.io/polkadot/book/runtime-api/index.html). In polkadot node this is wrapped in [runtime api subsystem](https://github.com/paritytech/polkadot/tree/3646202110b3810027f9d7fd0b45cb1bcb9678f4/node/core/runtime-api).

What you see in this file (`primitives/src/runtime_api.rs`) is a definition of all runtime apis implemented in polkadot runtime. These apis are declared with `decl_runtime_apis!` macro from substrate. This happens [here](https://github.com/paritytech/polkadot/blob/906d9115c1158ca915fc09bebe1e11baec9e4a16/primitives/src/runtime_api.rs#L120). There are multiple runtiime apis declared in the macro - each api is a separate trait. Usually we are interested in [`ParachainHost`](https://github.com/paritytech/polkadot/blob/906d9115c1158ca915fc09bebe1e11baec9e4a16/primitives/src/runtime_api.rs#L123) api.

About `v2` and `vstaging`. Each api has got a version. At the moment `v2` is the stable one which is on production. However during development we want to add new methods in the api and we don't want these to go in the production runtime. This is why `vstaging` exists. It contains methods which are not considered stable at the moment and they should be defined only in development runtimes. Runtimes can have different version so think about staging as 'runtime version bigger than 2 (or the current stable one).

How I know v2 is the stable version?

Check [here](https://github.com/paritytech/polkadot/blob/906d9115c1158ca915fc09bebe1e11baec9e4a16/primitives/src/runtime_api.rs#L122). The attribute sets a default version for the runtime api. Then [here](https://github.com/paritytech/polkadot/blob/906d9115c1158ca915fc09bebe1e11baec9e4a16/primitives/src/runtime_api.rs#L223) you can see a staging method which is tagged with v3. All the primitives it uses were available in v2 so you don't see vstaging module used there. However if this method had to introduce a new primitive it would have been put in staging.

And one final thing - runtime version vs runtime api version. These are used almost interchangeably but they are different. There is a version of the runtime as a whole (I think it's something like 9200) and version of each runtime api in this runtime. These are different things, but generally we care only about the version of the api, not the version of the runtime itself.

There are more details about how runtime api versioning works in the beginning of the file ([link](https://github.com/paritytech/polkadot/blob/906d9115c1158ca915fc09bebe1e11baec9e4a16/primitives/src/runtime_api.rs#L17)). 

If you need any more details feel free to ask or ping me directly. I'll be more than happy to discuss this further.

Reply:
I appreciate the explanation @tdimitrov!